### ISSUE TITLE
GraphBLAS JIT configuration is broken on the Mac (for direct compilation)

### ISSUE THREAD

**Issue by DrTimothyAldenDavis** [DrTimothyAldenDavis/SuiteSparse#563]
@mmuetzel : I have a fix for this issue but I wanted to ask you what you think.

The change a few weeks ago to GraphBLAS/cmake_modules/GraphBLAS_JIT_configure.cmake broke the compilation of JIT kernels on my M1 MacBook with this PR:

https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/c697550beb30046c5078c318f2d7e58751b5b2fd

The problem is that -lomp doesn't work.  The installation of OpenMP on the Mac says that libomp.dylib shouldn't be added to the default system paths because it can break things.  The new cmake converts  /usr/local/lib/libomp.dylib to -lomp, but the compilation fails when using -lomp.  It works fine with /usr/local/lib/libomp.dylib.

This shows up in the wildtype_demo:

```
(iso wait:C (setElement:to non-iso) 0 zombies, 1 pending)  [ GrB_transpose (wait:A 0 zombies, 2 pending) (build) 
(jit: compile and load) (jit: sh -c "/Library/Developer/CommandLineTools/usr/bin/cc -DGB_JIT_RUNTIME=1  -O3 -DNDEBUG -Xclang -fopenmp  -fPIC  -arch arm64  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk  -I/Users/davis/.SuiteSparse/GrB8.3.0/src  -I/usr/local/include -o /Users/davis/.SuiteSparse/GrB8.3.0/c/de/GB_jit__build__02eeeee__wildtype.o -c /Users/davis/.SuiteSparse/GrB8.3.0/c/de/GB_jit__build__02eeeee__wildtype.c    ; 
/Library/Developer/CommandLineTools/usr/bin/cc  -O3 -DNDEBUG -Xclang -fopenmp  -fPIC  -arch arm64  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk   -dynamiclib  -o /Users/davis/.SuiteSparse/GrB8.3.0/lib/de/libGB_jit__build__02eeeee__wildtype.dylib 
/Users/davis/.SuiteSparse/GrB8.3.0/c/de/GB_jit__build__02eeeee__wildtype.o
  -lm -ldl -lomp   ") ld: library 'omp' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
(jit: compiler error; compilation disabled) (generic build) add:(H<.>=H+H) (jit: not compiled) (generic add: wildtype) (hyper to sparse) (transpose) (builder transpose) (jit: not compiled) (hyper to sparse) 
```

The demo still runs because of how I handle JIT errors:  if a JIT fails to compile, I punt to a generic kernel with function pointers. But this disables the JIT entirely.

If I cut-and-paste that command but revise it, and replace -lomp with /usr/local/lib/libomp.dylib, the kernel compiles.

If the JIT uses cmake, it works fine since GB_C_LIBRARIES is not used.  GB_CMAKE_LIBARIRES works fine.

If I revert that body of code, I get this and it compiles just fine:

```
(iso wait:C (setElement:to non-iso) 0 zombies, 1 pending)  [ GrB_transpose (wait:A 0 zombies, 2 pending) (build) 
(jit: compile and load) (jit: sh -c "/Library/Developer/CommandLineTools/usr/bin/cc -DGB_JIT_RUNTIME=1  -O3 -DNDEBUG -Xclang -fopenmp  -fPIC  -arch arm64  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk  -I/Users/davis/.SuiteSparse/GrB8.3.0/src  -I/usr/local/include -o /Users/davis/.SuiteSparse/GrB8.3.0/c/de/GB_jit__build__02eeeee__wildtype.o -c /Users/davis/.SuiteSparse/GrB8.3.0/c/de/GB_jit__build__02eeeee__wildtype.c    ; /Library/Developer/CommandLineTools/usr/bin/cc  -O3 -DNDEBUG -Xclang -fopenmp  -fPIC  -arch arm64  -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk   -dynamiclib  -o /Users/davis/.SuiteSparse/GrB8.3.0/lib/de/libGB_jit__build__02eeeee__wildtype.dylib 
/Users/davis/.SuiteSparse/GrB8.3.0/c/de/GB_jit__build__02eeeee__wildtype.o
  -lm -ldl /usr/local/lib/libomp.dylib   ")
```

I realize that the new configuration is required for MING, but it breaks the Mac.   As a workaround, I've done this:

https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/c4581154a6342eba8adf9282a388cfa8c4a595eb

which fixes the problem on my M1 MacBook, at least.  I haven't tested it everywhere yet but I can run the CI.  What do you think?  Is the relocation issue that you had to fix for MING with your PR a few weeks ago something that should be done for most platforms?  Or is my PR today OK, so the -lomp is used on MINGW, but /usr/local/lib/libgomp.so (or whatever) is used elsewhere?

**Comment by mmuetzel** [comment]
IIRC, that part of the rules isn't really MinGW-specific. But it allows to relocate the library after it was built.

Some distributors build their packages in a "clean" environment. That "clean" environment can have a different prefix than the one that will be used when the package is installed in the final distribution. By hardcoding the prefix for these libraries, relocation can become an issue.

Would your change work for distributions like Homebrew?

Maybe, it would be better to at least invert the logic and use a special-case only for macOS (where OpenMP seems to be different) but keep relocation working on all other platforms.

**Comment by DrTimothyAldenDavis** [comment]
That sounds like a better plan.  Linux is fine with this MINGW variant.

Homebrew is a bit odd because it doesn't put OpenMP in its standard location:

```
mac $ brew info libomp
==> libomp: stable 17.0.6 (bottled) [keg-only]
LLVM's OpenMP runtime library
https://openmp.llvm.org/
/opt/homebrew/Cellar/libomp/17.0.3 (7 files, 1.7MB)
  Poured from bottle using the formulae.brew.sh API on 2023-10-20 at 09:25:48
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/lib/libomp.rb
License: MIT
==> Dependencies
Build: cmake ✘, lit ✘
==> Caveats
libomp is keg-only, which means it was not symlinked into /opt/homebrew,
because it can override GCC headers and result in broken builds.
```

My libomp that I'm using actually comes from this source, however:

https://mac.r-project.org/openmp/

In either case, on the Mac, -lomp or -lgomp would likely never work, and an explicit path to the OpenMP library would be needed, I think.  That's because even Homebrew's OpenMP doesn't get put in the standard places.

I could reach out to the Homebrew distro folks and ask them what they think.  One option for them would be to simply install their own variant of GB_config.h with the list of libraries they want.

**Comment by DrTimothyAldenDavis** [comment]
I'll wait to revise this after I merge your PR in for all the cmake changes.

One solution would be to add a cmake option, which defaults to "keep C libraries as explicit, as in /usr/lib/libomp.dylib", or "change libs to -lomp to enable relocation".  Then the default would be to use /usr/lib/libomp.dylib on the Mac, and -lomp for Linux, MINGW, Windows, and all others.

**Comment by DrTimothyAldenDavis** [comment]
Ah .. 2 steps forward one step back.  CHOLMOD fails to compile.  It files /usr/local/include/amd.h (and others).

```
[  2%] Building C object CMakeFiles/CHOLMOD_static.dir/Cholesky/cholmod_l_amd.c.o
/Library/Developer/CommandLineTools/usr/bin/cc -DBLAS_Apple
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/build
-I/Users/davis/dev2/SuiteSparse/CHOLMOD
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/Check
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/Cholesky
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/Utility
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/MatrixOps
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/Modify
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/Partition
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/Supernodal
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/Include
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/SuiteSparse_metis/GKlib
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/SuiteSparse_metis/libmetis
-I/Users/davis/dev2/SuiteSparse/CHOLMOD/SuiteSparse_metis/include
-F/Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk/System/Library/Frameworks
-isystem /Users/davis/dev2/SuiteSparse/SuiteSparse_config
-isystem /usr/local/include
-isystem /Users/davis/dev2/SuiteSparse/AMD/Include
-isystem /Users/davis/dev2/SuiteSparse/COLAMD/Include
-isystem /Users/davis/dev2/SuiteSparse/CAMD/Include
-isystem /Users/davis/dev2/SuiteSparse/CCOLAMD/Include
-O3 -DNDEBUG -std=gnu11 -arch
arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk
-mmacosx-version-min=13.6 -Xclang -fopenmp -MD -MT
CMakeFiles/CHOLMOD_static.dir/Cholesky/cholmod_l_amd.c.o -MF
CMakeFiles/CHOLMOD_static.dir/Cholesky/cholmod_l_amd.c.o.d -o
CMakeFiles/CHOLMOD_static.dir/Cholesky/cholmod_l_amd.c.o -c
/Users/davis/dev2/SuiteSparse/CHOLMOD/Cholesky/cholmod_l_amd.c make[2]: ***
[CMakeFiles/CHOLMOD.dir/all] Error 2 make[2]: *** Waiting for unfinished
jobs....  In file included from
/Users/davis/dev2/SuiteSparse/CHOLMOD/Cholesky/cholmod_l_amd.c:12:
/Users/davis/dev2/SuiteSparse/CHOLMOD/Cholesky/cholmod_amd.c:37:2: error:
"CHOLMOD:Cholesky requires AMD 3.3.0 or later" #error "CHOLMOD:Cholesky
requires AMD 3.3.0 or later"
```

**Comment by DrTimothyAldenDavis** [comment]
This was a fresh build, as well.

**Comment by DrTimothyAldenDavis** [comment]
The build using the top-level root CMakeLists.txt file worked, at least for this file.

It failed at cholmod_di_demo, but that's because I haven't merged in my changes to the CHOLMOD/CMakeLists.txt yet.  The CHOLMOD demo programs now check the version of AMD, CAMD, etc in the header and the library, with new functions like amd_version, etc.  So those demos will need to link agains AMD, etc.  That's a simple fix.

**Comment by DrTimothyAldenDavis** [comment]
oops, wrong thread ...

**Comment by DrTimothyAldenDavis** [comment]
This is now fixed in SuiteSparse 7.4.0.beta1.
